include/swift/Reflection/ReflectionContext.h

```diff
@@ -710,6 +711,128 @@ class ReflectionContext
+
+    auto EnumRecordTI = dyn_cast<const RecordTypeInfo>(EnumTI);
+    if (EnumRecordTI == nullptr)
+      return false;
+    auto EnumSize = EnumRecordTI->getSize();
+
+    auto Fields = EnumRecordTI->getFields();
+    auto FieldCount = Fields.size();
+    if (FieldCount == 0) {
+      return false;  // No fields?
+    }
+    if (FieldCount == 1) {
+      *CaseIndex = 0; // Only possible field
+      return true;
+    }
+
+    switch (EnumRecordTI->getRecordKind()) {
+
+    case RecordKind::NoPayloadEnum: {
+      if (EnumSize == 0) {
+        *CaseIndex = 0;
+        return true;
+      }
+      return getReader().readInteger(EnumAddress, EnumSize, CaseIndex);
+    }
+
+    case RecordKind::SinglePayloadEnum: {
+      FieldInfo PayloadCase = Fields[0];
+      if (!PayloadCase.TR)
+        return false;
+      unsigned long NonPayloadCaseCount = FieldCount - 1;
+      unsigned long PayloadExtraInhabitants = PayloadCase.TI.getNumExtraInhabitants();
+      unsigned discriminator = 0;
+      auto PayloadSize = PayloadCase.TI.getSize();
+      if (NonPayloadCaseCount >= PayloadExtraInhabitants) {
+        // There are more cases than inhabitants, we need a separate discriminator.
+        auto TagInfo = getEnumTagCounts(PayloadSize, NonPayloadCaseCount, 1);
+        auto TagSize = TagInfo.numTagBytes;
+        auto TagAddress = RemoteAddress(EnumAddress.getAddressData() + PayloadSize);
+        if (!getReader().readInteger(TagAddress, TagSize, &discriminator))
+          return false;
+      }
+
+      if (PayloadExtraInhabitants == 0) {
+        // Payload has no XI, so discriminator fully determines the case
+        *CaseIndex = discriminator;
+        return true;
+      } else if (discriminator == 0) {
+        // The value overlays the payload ... ask the payload to decode it.
+        int t;
+        if (!PayloadCase.TI.readExtraInhabitantIndex(getReader(), EnumAddress, &t)) {
+          return false;
+        }
+        if (t < 0) {
+          *CaseIndex = 0;
+          return true;
+        } else if ((unsigned long)t <= NonPayloadCaseCount) {
+          *CaseIndex = t + 1;
+          return true;
+        }
+        return false;
+      } else {
+        // The entire payload area is available for additional cases:
+        auto TagSize = std::max(PayloadSize, 4U); // XXX TODO XXX CHECK THIS
+        auto offset = 1 + PayloadExtraInhabitants; // Cases coded with discriminator = 0
+        unsigned casesInPayload = 1 << (TagSize * 8U);
+        unsigned payloadCode;
+        if (!getReader().readInteger(EnumAddress, TagSize, &payloadCode))
+          return false;
+        *CaseIndex = offset + (discriminator - 1) * casesInPayload + payloadCode;
+        return true;
+      }
+    }
+
+    case RecordKind::MultiPayloadEnum: {
+      // TODO: Support multipayload enums
+      break;
+    }
+
+    default:
+      // Unknown record kind.
+      break;
+    }
+    return false;
+  }
+
+  bool getEnumCaseTypeRef(const TypeRef *EnumTR,
```

Comment:
I think this API is redundant. In the C API, we already have swift_reflection_childOfTypeRef()